pennylane/data/attributes/operator/operator.py

```diff
@@ -0,0 +1,264 @@
+# Copyright 2018-2023 Xanadu Quantum Technologies Inc.
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Contains AttributeType definition for pennylane operators, and lists
+of operators."""
+
+import json
+import typing
+from functools import lru_cache
+from typing import FrozenSet, Generic, List, Type, TypeVar
+
+import numpy as np
+
+import pennylane as qml
+from pennylane.data.base._hdf5 import h5py
+from pennylane.data.base.attribute import AttributeType
+from pennylane.data.base.typing_util import HDF5Group
+from pennylane.operation import Operator, Tensor
+
+from ._wires import wires_to_json
+
+Op = TypeVar("Op", bound=Operator)
+
+
+class DatasetOperator(Generic[Op], AttributeType[HDF5Group, Op, Op]):
+    """``AttributeType`` for ``pennylane.operation.Operator`` classes.
+
+    Supports all operator types that meet the following conditions:
+        - The ``__init__()`` method matches the signature of ``Operator.__init__``,
+            or any additional arguments are optional and do not affect the value of
+            the operator
+        - The ``data`` and ``wires`` attributes will produce an identical copy of
+            operator if passed into the classes' ``__init__()`` method. Generally,
+            this means ``__init__()`` do not mutate the ``params`` and ``wires``
+            arguments.
+        - Hyperparameters are not used or are automatically derived by ``__init__()``.
```

Comment:
Is there a particular difficulty in being able to serialize hyperparameters?  Would there be a representation for "non trainable information" that would allow it to be nicely serialized?

Reply:
Probably not, I hadn't considered the issue too closely since it we didn't need to for the current datasets. The only potential issue whether the the types of hyperparameters are generally constrained (e.g JSON types, numbers, 1D arrays), or they can be any Python object. The latter case will be more difficult